### ISSUE TITLE
Tail Call Optimisation

### ISSUE THREAD

**Issue by Dichloromethane** [JuliaLang/julia#53033]
Hi everyone,

I noticed that the julia compiler does not seem to do tail call optimisation:

```
julia> @noinline id(x) = x
id (generic function with 1 method)

julia> f(x) = id(x)
f (generic function with 1 method)

julia> @code_native f(1)
<...snip...>
julia_f_154:                            # 
<...snip...>
	push	rbp
	mov	rbp, rsp
	movabs	rax, offset j_id_156
	call	rax
	pop	rbp
	ret
<...snip...>
```

which could be emitted as (copying the very strange and pessimising habit of the compiler to emit function calls as a `movabs`/indirect call pair)

```
julia_f_154:
  	movabs	rax, offset j_id_156
        jmp rax
```

which has all sorts of obvious benefits.

PS: I hope I have not missed an already existing bug report again somewhere, if so, sorry!

**Comment by nsajko** [comment]
#4964 dup?

**Comment by nsajko** [comment]
Note that TCE may be required for good performance when you use mutual recursion, but in general it's not a good idea because it ruins stack traces.

**Comment by Dichloromethane** [comment]
Ah, I obviously did not expect it to be closed.

If people are worried about stack traces, this then seems like something to enable at -O3 or -g0, because even in cases without tail recursion, it might make quite a difference.

**Comment by Dichloromethane** [comment]
I just had a discussion with several other julia users on the discord, and the premise under which this was closed was universally seen as "unfortunate". 

The idea is that julia is supposed to be used in HPC contexts, and there, the performance gain from TCO in non-recursive cases can really matter (because inlining the functions would lead to excessive code bloat, and a few percent of performance can really matter there)

**Comment by vchuravy** [comment]
Closed as duplicate of #4964, discussion can and should continue there. Stefan's opinion in  https://github.com/JuliaLang/julia/issues/4964#issuecomment-961932928 is probably still accurate.
